pdns/recursordist/recursor_cache.cc

```diff
@@ -856,6 +895,375 @@ void MemRecursorCache::doPrune(time_t now, size_t keep)
   pruneMutexCollectionsVector<SequencedTag>(now, d_maps, keep, cacheSize);
 }
 
+enum class PBCacheDump : protozero::pbf_tag_type
+{
+  required_string_version = 1,
+  required_string_identity = 2,
+  required_uint64_protocolVersion = 3,
+  required_int64_time = 4,
+  required_string_type = 5,
+  repeated_message_cacheEntry = 6,
+};
+
+enum class PBCacheEntry : protozero::pbf_tag_type
+{
+  repeated_bytes_record = 1,
+  repeated_bytes_sig = 2,
+  repeated_message_authRecord = 3,
+  required_bytes_name = 4,
+  required_bytes_authZone = 5,
+  required_message_from = 6,
+  optional_bytes_netmask = 7,
+  optional_bytes_rtag = 8,
+  required_uint32_state = 9,
+  required_int64_ttd = 10,
+  required_uint32_orig_ttl = 11,
+  required_uint32_servedStale = 12,
+  required_uint32_qtype = 13,
+  required_bool_auth = 14,
+  required_bool_submitted = 15,
+  required_bool_tooBig = 16,
+};
+
+enum class PBComboAddress : protozero::pbf_tag_type
+{
+  required_uint32_port = 1,
+  required_bytes_address = 2, // family implicit
+};
+
+enum class PBAuthRecord : protozero::pbf_tag_type
+{
+  required_bytes_name = 1,
+  required_bytes_rdata = 2,
+  required_uint32_type = 3,
+  required_uint32_class = 4,
+  required_uint32_ttl = 5,
+  required_uint32_place = 6,
+  required_uint32_clen = 7,
+};
+
+template <typename T>
+static void encodeComboAddress(protozero::pbf_builder<T>& writer, T type, const ComboAddress& address)
+{
+  protozero::pbf_builder<PBComboAddress> message(writer, type);
+
+  // Skip all parts except address and port
+  message.add_uint32(PBComboAddress::required_uint32_port, address.getPort());
+  if (address.sin4.sin_family == AF_INET) {
+    message.add_bytes(PBComboAddress::required_bytes_address, reinterpret_cast<const char*>(&address.sin4.sin_addr.s_addr), sizeof(address.sin4.sin_addr.s_addr)); // NOLINT(cppcoreguidelines-pro-type-reinterpret-cast): it's the API
+  }
+  else if (address.sin4.sin_family == AF_INET6) {
+    message.add_bytes(PBComboAddress::required_bytes_address, reinterpret_cast<const char*>(&address.sin6.sin6_addr.s6_addr), sizeof(address.sin6.sin6_addr)); // NOLINT(cppcoreguidelines-pro-type-reinterpret-cast): it's the API
+  }
+  else {
+    assert(0);
+  }
+}
+
+template <typename T>
+static void decodeComboAddress(protozero::pbf_message<T>& reader, ComboAddress& address)
+{
+  address.reset();
+  protozero::pbf_message<PBComboAddress> message(reader.get_message());
+
+  // Skip all parts except address and port
+  if (message.next(PBComboAddress::required_uint32_port)) {
+    address.setPort(message.get_uint32());
+  }
+  else {
+    assert(0);
```

Comment:
`assert` seems a bit harsh, we could terminate the process because of a corrupted file.